### PR TITLE
Bug 1215436 - Unit test for SyncBrowserDB r=yifan

### PR DESCRIPTION
**Pull request by weilonge**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=3c8420ae34033327e8f9cee59a26a5562832424b

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=397d36163fa92cb141d0d5c5133660dfe9834bc0

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=eaea245bd5f7199a6f61afadb33c6c8116dc567c

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=60e1e74ea9c58f34eddb6678d189878ed3755223

File: tv_apps/browser/test/unit/sync/db_test.js [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+/* jshint maxlen:80 */
+const DEFAULT_FAVICON = 'style/images/default-fav.svg';
+
+requireApp('browser/js/sync/db.js');
+
+var clearBrowserStores = function(done) {
+  SyncBrowserDB.db.open(function() {
+    SyncBrowserDB.db.clearPlaces(function() {
```
Comment (begeeben) [inline tv_apps/browser/test/unit/sync/db_test.js]: How about
```
Promise.all([
  new Promise(SyncBrowserDB.db.clearPlaces),
  new Promise(SyncBrowserDB.db.clearVisits),
  new Promise(SyncBrowserDB.db.clearIcons),
  new Promise(SyncBrowserDB.db.clearBookmarks)
]).then(done);
```
?

File: tv_apps/browser/test/unit/sync/db_test.js [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+          }
+          done();
+        });
+      });
+    });
+
+    test('getAllBookmarkUris', function(done) {
+      const RECORD_SIZE = 10;
+      bookmarkGenerator(RECORD_SIZE, records => {
+        SyncBrowserDB.db.getAllBookmarkUris(bookmarks => {
```
Comment (begeeben) [inline tv_apps/browser/test/unit/sync/db_test.js]: Maybe rename ```bookmarks``` to ```bookmarkUris``` would make it easier to be distinguished?

File: tv_apps/browser/test/unit/sync/db_test.js [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+    test('deleteBookmark', function(done) {
+      var example = {
+        id: 'FXSYNCID_003',
+        bmkUri: 'http://mozilla.org/test3',
+        title: 'Mozilla',
+        parentid: 'FXS_P_ID_003',
+        timestamp: Date.now()
+      };
+      SyncBrowserDB.db.saveBookmark(example, () => {
+        SyncBrowserDB.db.deleteBookmark('FXSYNCID_003', () => {
```
Comment (begeeben) [inline tv_apps/browser/test/unit/sync/db_test.js]: Should we check if the ```example``` is correctly saved before deleting it?

File: tv_apps/browser/test/unit/sync/db_test.js [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+        });
+      });
+    });
+
+    test('saveIcon', function(done) {
+      var blob = new Blob(['hello', ' world'], {type: 'text/plain'});
+      var iconEntry = {
+        uri: 'http://mozilla.org/favicon.ico',
+        data: blob,
+        expiration: new Date().valueOf()
```
Comment (begeeben) [inline tv_apps/browser/test/unit/sync/db_test.js]: We could simply use ```Date.now()``` here, but it's OK as it is, no big deal.

File: tv_apps/browser/test/unit/sync/db_test.js [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+      };
+      SyncBrowserDB.db.saveIcon(iconEntry, () => {
+        done();
```
Comment (begeeben) [inline tv_apps/browser/test/unit/sync/db_test.js]: Should we add an assertion test here?

File: tv_apps/browser/test/unit/sync/db_test.js [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+    test('removeBookmark', function(done) {
+      var example = {
+        id: 'FXSYNCID_003',
+        bmkUri: 'http://mozilla.org/test3',
+        title: 'Mozilla',
+        parentid: 'FXS_P_ID_003',
+        timestamp: Date.now()
+      };
+      SyncBrowserDB.updateRawBookmark(example, () => {
+        SyncBrowserDB.removeBookmark(example.id, () => {
```
Comment (begeeben) [inline tv_apps/browser/test/unit/sync/db_test.js]: Should we check if the bookmark is updated successfully first?

File: tv_apps/browser/test/unit/sync/db_test.js [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+            assert.equal(bookmark, undefined);
+            done();
+          });
+        });
+      });
+    });
+
+    test('clearBookmarks', function(done) {
+      bookmarkGenerator(10, (rawRecords) => {
+        SyncBrowserDB.clearBookmarks().then(() => {
```
Comment (begeeben) [inline tv_apps/browser/test/unit/sync/db_test.js]: Same as above here. If the records are not saved, the test will still pass.

File: tv_apps/browser/test/unit/sync/db_test.js [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+
+    test('updateRawHistory & removeHistory', function(done) {
+      var example = {
+        fxsyncId: 'FXSYNCID_003',
+        uri: 'http://mozilla.org/test3',
+        title: 'Mozilla',
+        timestamp: Date.now()
+      };
+      SyncBrowserDB.updateRawHistory(example, () => {
+        SyncBrowserDB.removeHistory(example.uri, bookmark => {
```
Comment (begeeben) [inline tv_apps/browser/test/unit/sync/db_test.js]: Same as above.

File: tv_apps/browser/test/unit/sync/db_test.js [hunk 9]
```diff
@@ -0,0 +1,480 @@
...
+          assert.equal(bookmark, undefined);
+          done();
+        });
+      });
+    });
+
+    test('clearHistoryDeep', function(done) {
+      historyGenerator(10, (rawRecords) => {
+        SyncBrowserDB.clearHistoryDeep().then(() => {
```
Comment (begeeben) [inline tv_apps/browser/test/unit/sync/db_test.js]: Same as above.

File: tv_apps/browser/test/unit/sync/db_test.js [hunk 10]
```diff
@@ -0,0 +1,480 @@
...
+        SyncBrowserDB.getHistoryTimestamp(10070, 'prev', false, ts => {
+          assert.equal(ts, 10060);
+          done();
+        });
+      }, 10000);
+    });
+
+    test('getHistoryTimestamp with equal=true / dir=next', function(done) {
+      historyGenerator(10, (rawRecords) => {
+        SyncBrowserDB.getHistoryTimestamp(10075, 'next', true, ts => {
```
Comment (begeeben) [inline tv_apps/browser/test/unit/sync/db_test.js]: Should we change the timestamp from 10075 to 10080?

File: tv_apps/browser/test/unit/sync/db_test.js [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+          assert.equal(ts, 10080);
+          done();
+        });
+      }, 10000);
+    });
+
+    test('getHistoryTimestamp with equal=false / dir=next', function(done) {
+      historyGenerator(10, (rawRecords) => {
+        SyncBrowserDB.getHistoryTimestamp(10075, 'next', false, ts => {
```
Comment (begeeben) [inline tv_apps/browser/test/unit/sync/db_test.js]: Should we change the timestamp from 10075 to 10080?

File: tv_apps/browser/test/unit/sync/db_test.js [hunk 12]
```diff
@@ -0,0 +1,480 @@
...
+          assert.equal(ts, 10080);
```
Comment (begeeben) [inline tv_apps/browser/test/unit/sync/db_test.js]: Assert ```ts``` to equal ```10090```?

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=49db02e453254b7536916faba640ce415a1b4117

**Pull request merged**